From 1831b5e2f9ae64cf2edcbc6081660ba10182693e Mon Sep 17 00:00:00 2001
From: "JianFeng,Zhou" <jianfeng.zhou@intel.com>
Date: Fri, 31 Jul 2020 08:12:40 +0800
Subject: [PATCH] notify host by vsock after data partition erased

notify host by vsock after data partition erased, so host software can
execute secure erase
---
 install/wipe_data.cpp | 39 ++++++++++++++++++++++++++++++++++++++-
 1 file changed, 38 insertions(+), 1 deletion(-)

diff --git a/install/wipe_data.cpp b/install/wipe_data.cpp
index 765a815..244aaee 100644
--- a/install/wipe_data.cpp
+++ b/install/wipe_data.cpp
@@ -19,6 +19,8 @@
 #include <stdio.h>
 #include <string.h>
 #include <sys/stat.h>
+#include <sys/socket.h>
+#include <linux/vm_sockets.h>
 
 #include <functional>
 #include <vector>
@@ -101,9 +103,40 @@ bool WipeCache(RecoveryUI* ui, const std::function<bool()>& confirm_func) {
   return success;
 }
 
+#define HOST_VSOCK_PORT 2345
+#define CMD_ERASE_DATA  "erase-data"
+int notify_host(void)
+{
+  int fd;
+  struct sockaddr_vm sa = {
+    .svm_family = AF_VSOCK,
+    .svm_cid = VMADDR_CID_HOST,
+    .svm_port = HOST_VSOCK_PORT,
+  };
+
+  fd = socket(AF_VSOCK, SOCK_STREAM, 0);
+  if(fd < 0)
+    return - (2000 + errno);
+
+  if(connect(fd, (struct sockaddr*)&sa, sizeof(sa)) != 0) {
+    close(fd);
+    return - (3000 + errno);
+  }
+
+  int ret = send(fd, CMD_ERASE_DATA, strlen(CMD_ERASE_DATA), 0);
+  if(ret <= 0)
+    ret = - (4000 + errno);
+  else
+    ret = 0;
+
+  close(fd);
+  return ret;
+}
+
 bool WipeData(Device* device, bool convert_fbe) {
   RecoveryUI* ui = device->GetUI();
   ui->Print("\n-- Wiping data...\n");
+
   bool success = device->PreWipeData();
   if (success) {
     success &= EraseVolume(DATA_ROOT, ui, convert_fbe);
@@ -118,6 +151,10 @@ bool WipeData(Device* device, bool convert_fbe) {
   if (success) {
     success &= device->PostWipeData();
   }
+
+  int ret = notify_host();
+  ui->Print("inform_host return: %d\n", ret);
+
   ui->Print("Data wipe %s.\n", success ? "complete" : "failed");
   return success;
-}
\ No newline at end of file
+}
-- 
2.7.4

